### PR TITLE
feat(sql): breaking change💥 - change default behaviour of CREATE TABLE AS/INSERT INTO SELECT statements to be non-atomic and batched

### PR DESCRIPTION
**Pull request by nwoolmer**
Closes https://github.com/questdb/questdb/issues/4251.
Closes https://github.com/questdb/questdb/pull/4186.
This PR changes the default behaviour of the `CREATE TABLE AS`/`INSERT INTO` `SELECT` statements to be non-atomic by default.

This is a **BREAKING** change.

**Rationale**
[CREATE TABLE](https://questdb.io/docs/reference/sql/create-table/#create-table-as) and [INSERT](https://questdb.io/docs/reference/sql/insert/) are atomic by default.

A `SELECT` clause can be used with either of the statements to insert data into the table, which can be sourced from other tables, or generated.

These statements are frequently used to copy or migrate large amounts of data. Requiring this to be performed atomically can cause a large amount of data to be held in-memory, and the query may fail if there are insufficient resources to complete it.

Therefore, by default, `CREATE TABLE AS SELECT`/`INSERT INTO SELECT` statements will no longer be atomic, and will instead insert in batches.

If a failure occurs during a non-atomic insert, then the target table may contain partially inserted data. Therefore, one should use [deduplication](https://questdb.io/docs/concept/deduplication/) keys to enforce idempotency on the target table, so that failed inserts can be safely re-run. De-duplication is [fairly cheap](https://questdb.io/blog/solving-duplicate-data-performant-deduplication/) for normal workloads

**Migration notes**
- If you use `CREATE TABLE table_name AS SELECT` statements:
    - you may want to retain the old behaviour by using `ATOMIC` i.e `CREATE ATOMIC TABLE table_name AS ...`
    - you may want to override the default batch size (8192) by setting the global 
    `cairo.sql.create.table.model.batch.size` option in server.conf
    - you may want to override the default batch size for a single query using `CREATE BATCH n TABLE table_name AS ...`
- If you use `INSERT INTO table_name SELECT` statements:
    - you may want to retain the old behaviour by using `ATOMIC` i.e `INSERT ATOMIC INTO table_name SELECT ...`
    - you may want to override the default batch size (8192) by setting the global 
    `cairo.sql.insert.model.batch.size` option in server.conf
    - you may want to override the default batch size for a single query using `INSERT BATCH n INTO table_name SELECT ...` 

**Checklist**
- [x] Add configuration option to override default behaviour for `CREATE TABLE table_name AS SELECT` statements
    - `cairo.sql.create.table.model.batchsize= 8192`
- [x] Add configuration option to override default behaviour for `INSERT INTO table_name SELECT` statements
    -  `cairo.sql.insert.model.batchsize= 8192`
- [x] Modify SqlParser to add ATOMIC keyword to CREATE TABLE
- [x] Modify SqlParser to add ATOMIC keyword to INSERT 
- [x] Modify SqlParser to add BATCH keyword to CREATE TABLE
- [x] Ensure ATOMIC is only available for CREATE TABLE/INSERT INTO with SELECT
- [x] Ensure BATCH is only available for CREATE TABLE with SELECT  
- [x] Update tests to be compatible with new default
- [x] Update assertion output to include atomic/batch information 
- [x] Add tests for new config option
- [x] Add tests for new ATOMIC keyword
- [x] Add tests for new create table as BATCH behaviour
- [x] Add documented entry to production `server.conf` 
- [x] Update documentation to explain changes
- [x] Update console UI highlighting to highlight ATOMIC keyword.
- [x] Quantify appropriate default batch size and update configs (1 million)
- [x] Update `wal.apply.worker.yield.threshold` from 10ms to 1 second.

### CODE REVIEW

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 54 / 73 (73.97%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/griffin/model/InsertModel.java|0|1|00.00%|
|:large_blue_circle:|io/questdb/cairo/DefaultCairoConfiguration.java|1|2|50.00%|
|:large_blue_circle:|io/questdb/griffin/model/CreateTableModel.java|3|5|60.00%|
|:large_blue_circle:|io/questdb/cairo/CairoConfigurationWrapper.java|2|3|66.67%|
|:large_blue_circle:|io/questdb/griffin/SqlParser.java|33|47|70.21%|
|:large_blue_circle:|io/questdb/PropServerConfiguration.java|5|5|100.00%|
|:large_blue_circle:|io/questdb/PropertyKey.java|3|3|100.00%|
|:large_blue_circle:|io/questdb/griffin/SqlKeywords.java|7|7|100.00%|

**Comment by jerrinot** [comment]
fwiw: Postgres, MSSQL and probably others use the [`create with`](https://www.postgresql.org/docs/current/sql-createtableas.html) syntax:

In our case, it could look like this:
```sql
create table 
with (atomic=true)
as (select ..... )
timestamp(ts) partitioned by day;
```
This allows adding additional modifiers without introducing new keywords. I think it's worth considering.

**Comment by jerrinot** [comment]
Another consideration with new keywords: they have to be escaped when used
as identifiers. This means whenever we introduce a new keyword we are
breaking compatibility. A parser might star barking when a use uses the
newly introduced keyword  as eg. a table name.

“Perfection is achieved, not when there is nothing more to add, but when
there is nothing left to take away.”
Antoine de Saint Exupéry

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 82 / 89 (92.13%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/griffin/model/InsertModel.java|0|1|00.00%|
|:large_blue_circle:|io/questdb/cairo/DefaultCairoConfiguration.java|1|2|50.00%|
|:large_blue_circle:|io/questdb/cairo/CairoConfigurationWrapper.java|2|3|66.67%|
|:large_blue_circle:|io/questdb/griffin/model/CreateTableModel.java|15|16|93.75%|
|:large_blue_circle:|io/questdb/griffin/SqlParser.java|49|52|94.23%|
|:large_blue_circle:|io/questdb/PropServerConfiguration.java|5|5|100.00%|
|:large_blue_circle:|io/questdb/PropertyKey.java|3|3|100.00%|
|:large_blue_circle:|io/questdb/griffin/SqlKeywords.java|7|7|100.00%|

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 82 / 89 (92.13%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/griffin/model/InsertModel.java|0|1|00.00%|
|:large_blue_circle:|io/questdb/cairo/DefaultCairoConfiguration.java|1|2|50.00%|
|:large_blue_circle:|io/questdb/cairo/CairoConfigurationWrapper.java|2|3|66.67%|
|:large_blue_circle:|io/questdb/griffin/model/CreateTableModel.java|15|16|93.75%|
|:large_blue_circle:|io/questdb/griffin/SqlParser.java|49|52|94.23%|
|:large_blue_circle:|io/questdb/PropServerConfiguration.java|5|5|100.00%|
|:large_blue_circle:|io/questdb/PropertyKey.java|3|3|100.00%|
|:large_blue_circle:|io/questdb/griffin/SqlKeywords.java|7|7|100.00%|

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 82 / 89 (92.13%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/griffin/model/InsertModel.java|0|1|00.00%|
|:large_blue_circle:|io/questdb/cairo/DefaultCairoConfiguration.java|1|2|50.00%|
|:large_blue_circle:|io/questdb/cairo/CairoConfigurationWrapper.java|2|3|66.67%|
|:large_blue_circle:|io/questdb/griffin/model/CreateTableModel.java|15|16|93.75%|
|:large_blue_circle:|io/questdb/griffin/SqlParser.java|49|52|94.23%|
|:large_blue_circle:|io/questdb/PropServerConfiguration.java|5|5|100.00%|
|:large_blue_circle:|io/questdb/PropertyKey.java|3|3|100.00%|
|:large_blue_circle:|io/questdb/griffin/SqlKeywords.java|7|7|100.00%|

File: core/src/main/resources/io/questdb/site/conf/server.conf [hunk 1]
```diff
@@ -329,6 +332,9 @@ query.timeout.sec=60
 # size of InsertModel pool in SqlParser
 #cairo.sql.insert.model.pool.capacity=64
 
+# batch size of non-atomic inserts for INSERT INTO SELECT statements
+#cairo.sql.insert.model.batch.size=8192
```
Comment (bluestreak01) [inline core/src/main/resources/io/questdb/site/conf/server.conf:336]: This looks like very low number.

We typically not on the best code path when we're dealing with small transactions. 8k rows at a time falls into category of small. We should try to find optimal number. I will do some testing.

File: core/src/main/java/io/questdb/griffin/model/InsertModel.java [hunk 2]
```diff
@@ -134,6 +134,10 @@ public ExpressionNode getTableNameExpr() {
         return tableNameExpr;
     }
 
+    public boolean isAtomic() {
```
Comment (bluestreak01) [inline core/src/main/java/io/questdb/griffin/model/InsertModel.java:137]: unused

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 91 / 96 (94.79%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/cairo/DefaultCairoConfiguration.java|1|2|50.00%|
|:large_blue_circle:|io/questdb/cairo/CairoConfigurationWrapper.java|2|3|66.67%|
|:large_blue_circle:|io/questdb/griffin/SqlParser.java|48|51|94.12%|
|:large_blue_circle:|io/questdb/PropServerConfiguration.java|5|5|100.00%|
|:large_blue_circle:|io/questdb/PropertyKey.java|3|3|100.00%|
|:large_blue_circle:|io/questdb/griffin/SqlCompilerImpl.java|6|6|100.00%|
|:large_blue_circle:|io/questdb/griffin/model/CreateTableModel.java|19|19|100.00%|
|:large_blue_circle:|io/questdb/griffin/SqlKeywords.java|7|7|100.00%|

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 91 / 96 (94.79%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/cairo/DefaultCairoConfiguration.java|1|2|50.00%|
|:large_blue_circle:|io/questdb/cairo/CairoConfigurationWrapper.java|2|3|66.67%|
|:large_blue_circle:|io/questdb/griffin/SqlParser.java|48|51|94.12%|
|:large_blue_circle:|io/questdb/PropServerConfiguration.java|5|5|100.00%|
|:large_blue_circle:|io/questdb/PropertyKey.java|3|3|100.00%|
|:large_blue_circle:|io/questdb/griffin/SqlCompilerImpl.java|6|6|100.00%|
|:large_blue_circle:|io/questdb/griffin/model/CreateTableModel.java|19|19|100.00%|
|:large_blue_circle:|io/questdb/griffin/SqlKeywords.java|7|7|100.00%|

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 93 / 98 (94.90%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/cairo/DefaultCairoConfiguration.java|1|2|50.00%|
|:large_blue_circle:|io/questdb/cairo/CairoConfigurationWrapper.java|2|3|66.67%|
|:large_blue_circle:|io/questdb/griffin/SqlParser.java|48|51|94.12%|
|:large_blue_circle:|io/questdb/PropServerConfiguration.java|5|5|100.00%|
|:large_blue_circle:|io/questdb/PropertyKey.java|3|3|100.00%|
|:large_blue_circle:|io/questdb/griffin/SqlCompilerImpl.java|6|6|100.00%|
|:large_blue_circle:|io/questdb/griffin/model/CreateTableModel.java|21|21|100.00%|
|:large_blue_circle:|io/questdb/griffin/SqlKeywords.java|7|7|100.00%|

File: core/src/main/resources/io/questdb/site/conf/server.conf (hunk 1 above)
Comment (bluestreak01) [reply core/src/main/resources/io/questdb/site/conf/server.conf]: 1000000

File: core/src/main/java/io/questdb/griffin/SqlCompilerImpl.java [hunk 3]
```diff
@@ -3295,7 +3310,7 @@ private void backupTable(@NotNull TableToken tableToken) {
                         }
                         RecordCursor cursor = reader.getCursor();
                         //statement/query timeout value  is most likely too small for backup operation
-                        copyTableData(cursor, reader.getMetadata(), backupWriter, writerMetadata, recordToRowCopier, SqlExecutionCircuitBreaker.NOOP_CIRCUIT_BREAKER);
+                        copyTableData(cursor, reader.getMetadata(), backupWriter, writerMetadata, recordToRowCopier, -1, 0, SqlExecutionCircuitBreaker.NOOP_CIRCUIT_BREAKER);
```
Comment (bluestreak01) [inline core/src/main/java/io/questdb/griffin/SqlCompilerImpl.java:3313]: we should be using defaults for backup too, even though this backup method is legacy

File: core/src/main/java/io/questdb/griffin/SqlParser.java [hunk 4]
```diff
@@ -1551,8 +1599,12 @@ private ExecutionModel parseInsert(GenericLexer lexer, SqlParserCallback sqlPars
             if (SqlKeywords.isO3MaxLagKeyword(tok)) {
                 int pos = lexer.getPosition();
                 model.setO3MaxLag(SqlUtil.expectMicros(tok(lexer, "lag value"), pos));
-                expectTok(lexer, "into");
+                tok = tok(lexer, "into");
             }
+        } else if (SqlKeywords.isIntoKeyword(tok)) {
```
Comment (bluestreak01) [inline core/src/main/java/io/questdb/griffin/SqlParser.java:1604]: please remove empty conditional

File: core/src/test/java/io/questdb/test/griffin/CreateTableAsSelectTest.java [hunk 5]
```diff
@@ -87,4 +132,63 @@ private void testCreatePartitionedTableAsSelectWithOrderBy(String orderByClause)
             );
         });
     }
+
+    private void createPartitionedTableAsSelectWithOrderBy(String orderByClause, int batchSize, String o3MaxLag) throws Exception {
```
Comment (bluestreak01) [inline core/src/test/java/io/questdb/test/griffin/CreateTableAsSelectTest.java:136]: batchSize is constant expression, there is warning for it. Please use variety of values in tests

**Review by bluestreak01 — CHANGES_REQUESTED** [review]

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 103 / 108 (95.37%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/cairo/CairoConfigurationWrapper.java|2|3|66.67%|
|:large_blue_circle:|io/questdb/cairo/DefaultCairoConfiguration.java|4|5|80.00%|
|:large_blue_circle:|io/questdb/griffin/SqlParser.java|45|48|93.75%|
|:large_blue_circle:|io/questdb/PropServerConfiguration.java|9|9|100.00%|
|:large_blue_circle:|io/questdb/PropertyKey.java|3|3|100.00%|
|:large_blue_circle:|io/questdb/griffin/SqlCompilerImpl.java|11|11|100.00%|
|:large_blue_circle:|io/questdb/griffin/model/CreateTableModel.java|22|22|100.00%|
|:large_blue_circle:|io/questdb/griffin/SqlKeywords.java|7|7|100.00%|

**Review by bluestreak01 — APPROVED** [review]

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 103 / 108 (95.37%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/cairo/CairoConfigurationWrapper.java|2|3|66.67%|
|:large_blue_circle:|io/questdb/cairo/DefaultCairoConfiguration.java|4|5|80.00%|
|:large_blue_circle:|io/questdb/griffin/SqlParser.java|45|48|93.75%|
|:large_blue_circle:|io/questdb/PropServerConfiguration.java|9|9|100.00%|
|:large_blue_circle:|io/questdb/PropertyKey.java|3|3|100.00%|
|:large_blue_circle:|io/questdb/griffin/SqlCompilerImpl.java|11|11|100.00%|
|:large_blue_circle:|io/questdb/griffin/model/CreateTableModel.java|22|22|100.00%|
|:large_blue_circle:|io/questdb/griffin/SqlKeywords.java|7|7|100.00%|

**Pull request merged**